[DMT]-1194 Implement Clustering

peterkemp-routific opened the pull request:
[Jira](https://routific.atlassian.net/browse/DMT-1194)

### Overview
When there are multiple items that are close together, and to tiny to interact with, we want to cluster them together so we can interact with that cluster. This is mainly used for timelines, such as delivery stops along a route.

By clustering these items, it increases performance by decreasing the number of items on the screen that are rendered.

This PR contains the following:

1.)  Throttling zoom to 100ms to increase performance.
2.) Added clustering demo page.
3.) Added clustering on the timeline via settings passed in.
4.) Sorting Items if not pre-sorted. (Added the option to pass into the timeline if items or sorted or not to save us double sorting).

### How the clustering works.

The clustering does the following:

- From the left, check each item. If an item is a tiny item, check to the left and right of it to see if there are any items that are within the clustering range.
- If there is an item, start clustering to the closest item. (Size does not matter).
- Move to the next item and check to see if that item is a tinyItem and within clustering range. If so, add it to the cluster and continue this logic until it reaches an item that does not.

You can configure the settings of clustering with the following options when passing into the timeline.

   clusterSettings={{
          tinyItemSize: 0.4,
          clusteringRange: 0.5,
          sequencialClusterTinyItemsOnly: true, 
        }}

- tinyItemSize:  The size of the item
- clusteringRange: how far between the current item and the next item in order to cluster.
- sequencialClusterTinyItemsOnly:  After the initial tinyItem has a cluster to it's nearest neighbour, you can set this to false so it will cluster with any item within clustering range, not just a tinyItem). *Good for performance at larger numbers as it creates more clusters, which equals less items on the dom to render.  



### Items left to do in this PR

- Make sure clustering supports immutable.js 
- Update Readme
- Update Changelog
- Disable clustering at certain zoom levels via settings
- Add component tests for clustering

### Work to be done in upcoming PR's

**Lint Upgrade**
-> Perform Linting Upgrade

**Default Render**
- Create default cluster render
- Create cluster functional component to handle interactions with a cluster.

- Perform Linting Upgrade
-
### Nice To Have
- Github Actions for Validation Checks
- -> Sonar
- -> EsLint: *Current linting fails and will need to update library.
- -> Tests

JustinmClapperton commented on src/lib/utility/Cluster.js:
```diff
@@ -0,0 +1,64 @@
+      return this.#startTime();
+    }
+
+    get itemsInCluster() {
```
Can probably just be `get items()` here since its called on a Cluster object

JustinmClapperton commented on src/lib/utility/Cluster.js:
```diff
@@ -0,0 +1,64 @@
+
+    #clusterHasRequiredFields() {
+      if (!this.#items) {
+        throw new Error('Please add items to your cluster');
```
We couldnt also maybe fail this case silently by just ignoring empty clusters vs throwing an error.

peterkemp-routific replied:
By the time someone gets to this state, they have created a cluster with a group id and have tried to call .toObject() on it. If they forgot to add an item to the cluster, I would rather throw an error vs returning an empty object or undefined vs having wonder why they are missing a cluster somewhere on their screen.  Maybe they tried to add an item, but the item is null or undefined.

peterkemp-routific committed:
Updating Clustering to allow immutable.js

Used generic methods for accessing object by index and getting the length.

peterkemp-routific committed:
Update Version and Linting

peterkemp-routific committed:
Added ability to change clustering to tiny stops only from settings being passed in

also updated some code smells

peterkemp-routific committed:
Added default cluster renderer, user supplied cluster renderer, throttle via props.

-> Added default cluster renderer.
-> Added ability to pass in a item renderer for a cluster to the timeline.
-> Added disable clustering when canvas duration is below a certain duration. (in ms).
-> Added ability to control the zoom throttling via props
-> Updated changelog features.

peterkemp-routific committed:
Component Tests

Adding initial clustering component tests.

Note: index.js was disabled before as there are previous failing tests. This is where the itemRenderClustering would normal go. During the library cleanup, we should look at why these tests are disabled and add in the rendering for clustering there with the other custom item renderer.

peterkemp-routific committed:
Merge branch 'master' into DMT-1194-implement-clustering

peterkemp-routific committed:
Merge branch 'DMT-1194-implement-clustering' into DMT-1650-clustering-component-tests

peterkemp-routific committed:
Merge branch 'DMT-1194-implement-clustering' into DMT-1652-default-cluster-renderer

peterkemp-routific committed:
Merge pull request #9 from routific/DMT-1652-default-cluster-renderer

[DMT-1652] - Added default cluster renderer, user supplied cluster renderer, throt…

peterkemp-routific committed:
Merge branch 'DMT-1194-implement-clustering' into DMT-1650-clustering-component-tests

peterkemp-routific committed:
Merge pull request #10 from routific/DMT-1650-clustering-component-tests

[DMT]-1650 clustering component tests
